packages/dev/core/test/unit/DeviceInput/babylon.webDeviceInputSystem.test.ts

```diff
@@ -0,0 +1,142 @@
+/**
+ * @jest-environment jsdom
+ */
+
+jest.mock("core/Misc/logger");
+
+import { WebDeviceInputSystem } from "core/DeviceInput/webDeviceInputSystem";
+import { NullEngine, NullEngineOptions } from "core/Engines/nullEngine";
+import { Logger } from "core/Misc/logger";
+
+const mockTraceWarn = Logger.Warn as jest.Mock<any, any>;
+
+describe("WebDeviceInputSystem", () => {
+    let engine: NullEngine;
+    let wdis: WebDeviceInputSystem;
+    let mockOnInputChanged: jest.Mock<any, any>;
+    let mockOnDeviceConnected: jest.Mock<any, any>;
+    let mockOnDeviceDisconnected: jest.Mock<any, any>;
+    let renderElement: HTMLCanvasElement;
+    let addEventListenerSpy: jest.SpyInstance;
+    let raiseOnPointerDown: (evt: any) => {};
+    let raiseOnPointerMove: (evt: any) => {};
+    let raiseOnPointerUp: (evt: any) => {};
+
+    beforeEach(() => {
+        // So that GetPointerPrefix knows we are going to simulate pointer events
+        window.PointerEvent = jest.fn() as any;
+
+        renderElement = document.createElement("canvas");
+        addEventListenerSpy = jest.spyOn(renderElement, "addEventListener");
+        addEventListenerSpy.mockImplementation((type, listener) => {
+            switch (type) {
+                case "pointerdown":
+                    raiseOnPointerDown = listener;
+                    break;
+                case "pointermove":
+                    raiseOnPointerMove = listener;
+                    break;
+                case "pointerup":
+                    raiseOnPointerUp = listener;
+                    break;
+            }
+        });
+
+        mockOnInputChanged = jest.fn();
+        mockOnDeviceConnected = jest.fn();
+        mockOnDeviceDisconnected = jest.fn();
+
+        const nullEngineOptions = new NullEngineOptions();
+        nullEngineOptions.renderingCanvas = renderElement;
+        engine = new NullEngine(nullEngineOptions);
+        wdis = new WebDeviceInputSystem(engine, mockOnDeviceConnected, mockOnDeviceDisconnected, mockOnInputChanged);
+    });
+
+    afterEach(() => {
+        engine.dispose();
+        wdis.dispose();

```

Comment:
Conceptually a device input system instance should have a lifetime shorter than the engine. I don't know if this is enforced in any way, but to be safe maybe do the teardown in the opposite order as the setup (e.g. swap the order of these two dispose calls).

Reply:
Agreed, swapped the order!